src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,407 @@
+            this.useGraph = true;
+        } else {
+            this.useGraph = false;
+        }
+
+        if (backEndType.equals(BackEndType.LOGIQL)) {
+            // Configure solving strategy.
+            this.solveInParallel = false;
+        } else if (solveInParallel == null || solveInParallel.equals(Constants.TRUE)) {
+            this.solveInParallel = true;
+        } else {
+            this.solveInParallel = false;
+        }
+
+        if (collectStatistic == null || collectStatistic.equals(Constants.FALSE)) {
+            // Configure statistic collection.
+            this.collectStatistic = false;
+        } else if (collectStatistic.equals(Constants.TRUE)) {
+            this.collectStatistic = true;
+        }
+
+        // Sanitize the configuration if it needs.
+        sanitizeConfiguration();
+        System.out.println("Configuration: \nback end type: " + this.backEndType.simpleName + "; \nuseGraph: "
+                + this.useGraph + "; \nsolveInParallel: " + this.solveInParallel + ".");
+    }
+
+    protected void configureLattice(QualifierHierarchy qualHierarchy) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        lattice = latticeBuilder.buildLattice(qualHierarchy);
+    }
+
+    protected TwoQualifiersLattice createTwoQualifierLattice(AnnotationMirror top, AnnotationMirror bottom) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        return latticeBuilder.buildTwoTypeLattice(top, bottom);
+    }
+
+    /**
+     * This method creates a ConstraintSerializer, which can deliver the
+     * constraint/slot to the real serializer. If a costume serialization logic
+     * is needed, user can have a subclass of ConstraintSerializer, override the
+     * serialize method with the costume logic, and then override this method
+     * with returning the instance of new subclass.
+     * 
+     * @param backEndType
+     * @param lattice
+     * @return A deliverer Serializer.
+     */
+    protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
+        return new ConstraintSerializer<>(backEndType, lattice);
+    }
+
+    protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Constraint> constraints,
+            ProcessingEnvironment processingEnvironment) {
+        GraphBuilder graphBuilder = new GraphBuilder(slots, constraints);
+        ConstraintGraph constraintGraph = graphBuilder.buildGraph();
+        return constraintGraph;
+    }
+
+    protected BackEnd<?, ?> createBackEnd(BackEndType backEndType, Map<String, String> configuration,
+            Collection<Slot> slots, Collection<Constraint> constraints,
+            QualifierHierarchy qualHierarchy, ProcessingEnvironment processingEnvironment,
+            Lattice lattice, Serializer<?, ?> defaultSerializer) {
+
+        BackEnd<?, ?> backEnd = null;
+
+        try {
+            Class<?> backEndClass = Class.forName(backEndType.fullyQualifiedName + "BackEnd");
```

Comment:
Okey, this makes sense to me, and I cannot think of other reasons to keep `ConstraintSerializer`. But what's the advantage to move the serializer creation logic to `BackEndType`? Can we just create the real serializer in `GeneralSolver`